### DISCUSSION TITLE
[Feature] On demand thumbnail and transcoding

### DISCUSSION THREAD

**Discussion by grmrgecko** [immich-app/immich#7666] [Feature Request]
### The feature

Hello,

I'd like to see an option to disable thumbnail and transcode creation and to see support for on demand thumbnail creation. I run immich on a 32 core threadripper with an RTX 3080 GPU, so transcoding in real time isn't difficult for my system.

I have searched around, and the closest discussion similar to this is https://github.com/immich-app/immich/discussions/1334 however my request is that we save disk space by not transcoding and creating thumbnails, but doing so in real time. Maybe the thumbs and encode-video folder can be periodically flushed of files older than a day, that way we can use those folders to temporarily store the needed thumbnail and transcode in a way that's similar to cache in that it gets cleared over time.

The reason I'm wondering about this is currently both of these folders add up to about 400 GB, which is not insignificant even though I have 25 TB free disk space.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by mmomjian** [comment]
Are you suggesting that the small WebP (scroll view) thumbnails would also be generated on demand? Have you considered the I/O requirements to read in that many images in full resolution when scrolling through the timeline? Or is this only for the "mid-size" JPEGs? I guess my point is that there are more things to think about than just raw CPU and GPU power, especially for pictures - video transcoding is a much more doable because videos are viewed at a fixed speed.

**Reply by grmrgecko** [reply]
That is definitely something that can be considered in a feature like this, maybe allow the option of small thumbnails to be kept around but the larger previews to on demand. I am not too concerned about it with my hardware, but I can see how it may be a concern for some people.

**Comment by FoxxMD** [comment]
There's also this issue asking for hardware-accelerated decoding https://github.com/immich-app/immich/issues/4321

**Comment by AngelaDMerkel** [comment]
Having the option to transcode video on the fly is probably where this feature would be most useful. The (tiny) still image and video thumbnails are hardly a storage concern, but even 720p video adds up. The problem is that this can only be realistically accomplished with hardware decoding and NVENC, QuickSync, etc are much much less robust and flexible than ffmpeg using x265 for example. I sometimes add video in strange resolutions or uncommon bit depths (12bit 4:4:4) which break hardware encoders. It seems like it would be very very tough to account for all the weirdness without having a whitelist for video which should be converted on the fly and which should be pretranscoded in software.

**Reply by Hadeur** [reply]
I just wanted to add: I recently discovered Immich and set it up on my NAS and love it so far. Since one of my main reasons to search for a Synology Photos alternative was the removal of its transcoding ability i'd really like to have the option to use on-the-fly transcoding for videos, maybe even when i use the Immich app on a mobile network.

**Reply by AngelaDMerkel** [reply]
Having almost another year of Immich since this comment, I am convinced this feature would be useful. I moved to a new NAS and imported all my images and videos into a new instance of Immich and the processing takes around a day for the videos, especially when the originals are 8K. I have not and am unlikely to watch a lot of these videos, so they don't need to be transcoded in advance, and if they could be hardware transcoded whenever I go to watch them, that would be satisfactory.

**Comment by dronnikovigor** [comment]
Well, now there are a lot of options to specify what to trancode and what not. Also there is an option to enable/disable hardware transcoding. I don't see a problem to add option to use on-fly transcoding or not.

**Reply by bo0tzz** [reply]
That's a gross misunderstanding, on-the-fly transcoding isn't just a line we add to the settings to have everything magically work, it's a complicated feature that would take a lot of effort to implement.

**Reply by dronnikovigor** [reply]
Sure it is. I just wanted to point, that feature may be optional as now it's for hardware transcoding (e.g. I keep it disabled as it's stuck at HDR videos) for those, who have problems with specific videos.

**Comment by AlmightyFrog** [comment]
I see no real point in having thumbnails generated on the fly, it would be instead worth to have it kind of be able to detect moved/duplicate files and not generate a new thumbnail again cluttering up more disk space than necessary.

When it comes to the video I instead see a real use for on the fly transcoding: having all those transcoded video files adding up does not make much sense for a mostly single user system where even my simple i5 with quicksync could enode on the fly...

**Reply by AngelaDMerkel** [reply]
I agree that this is the scenario where this would make sense, but getting around all the systems which don't have capable hardware transcoding would be a nuisance

**Comment by Nodiink** [comment]
I've been experimenting with Immich for a while now and I have to say that I'm quite happy with it. 
When I found the settings for HW encoding I honestly assumed that it was on demand, also because encoding via GPU produces larger files, although it takes much less time. The doubt came to me at the two-pass parameter, which didn't make sense with my assumption.
In my opinion, on demand encoding should be a must-have, and for those who can only offer the CPU the existing approach is more than valid, but for those who can offer the GPU, on demand transcoding has no cons and several pros, including the ability to change only the container while keeping the original tracks, or convert only one of them, and to convert based on the client and the network speed.

Clearly, implementing this feature, which impacts both frontend and backend, is far from simple: two of the four famous media servers currently available have it as a paid key feature, and for a reason.

While browsing, I noticed that Immich seems to use the ffmpeg library compiled by Jellyfin, and I thought of a possible solution. 
I'll be very abstract because at the moment I have no idea what the limitations are and therefore if it is really possible to do it or not, but ... have you considered the possibility of using Jellyfin itself, via a plugin, for transcoding? 
In my head I see two options:
- Immich offers to Jellyfin via its plugin the original file, and the plugin returns a transcode that Immich then offers to the client (the video would take a long path, but maybe it is easier to implement?);
- or that jellyfin itself is started, like a third-party app to manage the video file (such as when you are on a cloud app, and a third-party app opens the actual file) so if you are on a browser it opens the jellyfin url, if you use the app it opens the jellyfin app (or asks you to install it);

Again, I have no idea if it is possible and I think the feasibility is mostly linked to the limitations of the jellyfin plugins, I hope someone can be more helpful in this regard.

**Reply by bo0tzz** [reply]
AFAIK Jellyfin doesn't expose anything that would allow it to be used in this way. A similar project, [Kyoo](https://github.com/zoriya/Kyoo/), does and it has crossed my mind to use its transcoding backend (https://github.com/zoriya/Kyoo/tree/master/transcoder) like this before.

**Reply by Swell61** [reply]
On-the-fly video transcoding would be a killer feature for me too. I've been using the Nextcloud app Memories, which I'm trying to replace with Immich, and it has on-the-fly transcoding. Perhaps their implementation may offer some ideas?

**Comment by dionorgua** [comment]
On fly transcoding for 'playback' is surely the best thing to have. 

But as temporary it would be cool to have something like lazy on-demand transcoding. So that system will generate these video files only when actually needed and cache results just for some time (days/months/years depending on settings). User will need to wait until it's done to start playback. 

This will add way to not store tons of transcoded videos for old content that may be not played at all..

**Reply by AngelaDMerkel** [reply]
This is a great compromise and would also avoid reencoding a newer video if it's being routinely viewed. I'd like to see a cache duration though, so that older videos can be cleared from the cache. Assuming I used Immich for 5 or 10 years, it adds up.

**Comment by skatsubo** [comment]
Ongoing work:
- [draft: VOD tanscoding implementation #18022](https://github.com/immich-app/immich/pull/18022)
- [Real-time transcoding/streaming](https://discord.com/channels/979116623879368755/1327726696421523486/1327726696421523486) - its technical discussion on Discord

**Comment by ymj0424** [comment]
Is there an update to this feature? I really need this function. Although I have set the maximum bitrate for transcoding to 20Mbps, the current implementation of video playback in immich still cannot play smoothly with a 100Mbps upload bandwidth.

**Reply by thegabriele97** [reply]
The PR is closed 😔 no one is working on it AFIK

**Comment by Brewsk11** [comment]
Reposting my thoughts from #28075

---
Hi,

I'd like to preface this discussion that it is not a new feature request. What it is, is an attempt to raise priority and highlight community feedback on the live transcoding debate.

TLDR: I'd love Immich to acknowledge the need of and add a "video quality/bitrate selection in the video player" on the roadmap.

The use case I'm unhappy about is viewing and sharing uploaded videos. This is half of the main feature of Immich (the second being photos, which is excellent). Currently, the design decisions made around video playback, in my opinion, are dubious.
I get the "just works" philosophy: transcode videos ahead of time to a well supported format, ready to play whenever on whatever hardware. Maybe this is the best approach for some deployments, but in my usecases it is always an unfavourable tradeoff.

If I want to be able to play my videos anywhere, I must have an option to reduce the bandwidth requirement. When I'm with my friends outdoors with low mobile network signal I can't show them my diving trip, because 1080p won't load. When I'm at home on a 4K monitor the 1080p playback with reduced bitrate isn't optimal, is it?

Not in scope for this discussion, but closely related, is the original video playback issue. At home I can't watch the 4K playback either, because the video does not stream; instead, Immich needs to download it before playback (sometimes? https://github.com/immich-app/immich/issues/23530), which for 10GB+ files it is unreasonable. 

I have not used all of the personal media server solutions, but Nextcloud Memories does the live transcoding, and except some UI issues on mobile it works well, while being a much smaller project.

I'm being nonchalant about it, but to me it is _that_ important for the user experience, that without the live transcoding the video playback, to me, is barely functional. Every time I use Immich to view videos it's either too low quality to watch or too high for remote networks (or the original just won't play).

I'm not alone, here are some of the live transcoding requests in the community:

- https://github.com/immich-app/immich/discussions/7666
- https://github.com/immich-app/immich/discussions/10993#discussioncomment-10012209
- https://www.reddit.com/r/immich/comments/17mpqcb/live_transcoding/

Two years ago a maintainer posted this comment: https://github.com/immich-app/immich/discussions/10993#discussioncomment-10008504.

After two years Immich evolved a lot. It seems both community and project maintainers agree this is a good idea. To some, including me, essential. Yet, besides a couple issues/discussions, it has not materialized in plans.

The feature is obviously complex. I have faith the Immich feature backlog is prioritized well. I understand there probably are infrastructural changes far more important for user experience than live transcoding, but the roadmap also lists planned features like "Smart memories" (questionably a priority), "Workflows" (a complex feature probably mainly for power users), "Better sharing" (improving what already works).

So the question is: what is the maintainers/project owners stance on the video live transcoding? Is it already planned? Will its priority be reconsidered?

**Reply by mertalev** [reply]
As it happens, live transcoding is something we're working on right now. It's definitely a more flexible approach to playback since playback can be aware of bandwidth, video compatibility, HDR, etc. By contrast, the current offline transcoding requires you to decide upfront which clients and network conditions to support and hope this will never change over time.

There are genuine tradeoffs with real-time transcoding, though, with start latency probably being the biggest one. Doing things completely live adds a tangible delay, especially noticeable when hovering over a video or swiping between videos. The current approach is to pre-transcode the first 2-4s of playback so the video can start playing immediately while the real-time transcode has time to ramp up. It improves UX significantly at the expense of a bit of storage per video.

A related challenge is ABR: HLS clients see a video taking a while to download and just assume poor bandwidth, unaware that the server is still preparing the video. Pre-transcoding helps a lot here as well since the client can get a more accurate bandwidth estimate from the fast initial segment(s) and the client has enough playback buffer to not panic when the first live segment takes a bit to come.

This is all assuming playback starts at the start of the video, but playback after seeking can't be helped much. The solution there would probably be to lean more on trick play (not implemented yet).

Overall, there are many layers to this feature, without even getting into the technical details of video playlist and segment management. It will probably be an alpha/experimental feature at first so we can land it sooner and refine it over time.

**Reply by Brewsk11** [reply]
Thank you for the update, much appreciated!
